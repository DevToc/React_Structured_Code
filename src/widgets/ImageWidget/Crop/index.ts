export * from './Crop';
export * from './Crop.types';
export * from './Crop.helpers';
