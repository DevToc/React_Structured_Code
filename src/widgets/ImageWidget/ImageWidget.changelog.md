# ImageWidget Changelog

### Version 2

- Opacity option added into the ImageWidget toolbar.
- The opacity impacts the current Image's rect, same as frame and crop.
- Opacity will be stored as a number with a range between 0 and 1, but we show the option as percentage, so in the `OpacityOption.tsx` we multiply it by 100, but inside `ImageWidget.tsx` we show it as 0-1.


