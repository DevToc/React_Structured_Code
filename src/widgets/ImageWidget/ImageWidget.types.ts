import React from 'react';
import { Widget, AccessibleElement } from 'types/widget.types';
import { WidgetId } from 'types/idTypes';
import { FrameShape } from './Frame/frame.types';

export interface Crop {
  top: number;
  right: number;
  bottom: number;
  left: number;
}

export interface ImageRect {
  widthPx: number;
  heightPx: number;
}

export interface ImageWidgetData extends Widget, AccessibleElement {
  url: string;
  crop: Crop;
  imageRect: ImageRect;
  frame: FrameShape;
  // Opacity of the image widget, ranges between 0 - 1.
  opacity: number;
  // TODO: filters
  // TODO: link
}

export interface ImageLayoutProps {
  alt?: string;
  imageRectRef?: React.RefObject<HTMLDivElement>;
  imageRef?: React.RefObject<HTMLImageElement>;
}
