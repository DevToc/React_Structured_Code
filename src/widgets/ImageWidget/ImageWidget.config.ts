import { HANDLE } from 'constants/bounding-box';
import { Widget } from 'types/widget.types';
import { ImageWidgetData } from './ImageWidget.types';
import { isSquareFrame } from './Frame/frame';

export const BOUNDING_BOX_CONFIG = {
  customHandle: (activeWidget: Widget) => {
    const imageWidgetData = activeWidget as ImageWidgetData;
    if (isSquareFrame(imageWidgetData.frame)) return HANDLE.CORNERS;
    return HANDLE.ALL;
  },
};

export const CROP_BUTTON_ID = 'crop-button';
