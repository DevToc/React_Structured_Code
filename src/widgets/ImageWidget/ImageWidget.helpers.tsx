import { useEffect, useState } from 'react';
import { WidgetType } from 'types/widget.types';
import { calculateAspectRatioFit } from 'utils/calculateAspectRatioFit';
import { Crop } from './ImageWidget.types';
import { FrameShape } from './Frame/frame.types';
import { VERSION } from './ImageWidget.upgrade';
import { parseStrictNumber } from 'utils/number';

interface ImageLoader {
  isLoading: boolean;
}

/**
 * A Simple hook for getting the loading state of an image url
 * e.g. const { isLoading } = useImageLoader('cat.png')
 *
 * @param url
 * @returns {ImageLoader}
 */

const loadedUrls: { [url: string]: boolean } = {};

export const useImageLoader = (url: string): ImageLoader => {
  const [isLoading, setIsLoading] = useState(!loadedUrls[url]);

  useEffect(() => {
    if (loadedUrls[url]) return;
    if (!url) setIsLoading(false);

    const img = document.createElement('img');
    img.src = url;
    img.onload = () => {
      setIsLoading(false);
      loadedUrls[url] = true;
    };

    return () => {
      if (img) img.onload = () => {};
    };
  }, [url]);

  return { isLoading };
};

interface Dimension {
  width: number;
  height: number;
}

export const generateDefaultData = (
  url: string,
  imageSize: Dimension,
  infographSize: Dimension,
  topPxOverride?: number,
  leftPxOverride?: number,
) => {
  const { width: widthPx, height: heightPx } = calculateAspectRatioFit(
    imageSize.width,
    imageSize.height,
    // The new image widget is limited to max half the size of the page
    infographSize.width / 2,
    infographSize.height / 2,
  );

  const imageWidget = {
    widgetType: WidgetType.Image,
    widgetData: {
      version: VERSION,
      topPx: topPxOverride || 0,
      leftPx: leftPxOverride || 0,
      rotateDeg: 0,
      widthPx,
      heightPx,
      url: url,
      altText: '',
      isDecorative: false,
      crop: {
        top: 0,
        right: 0,
        bottom: 0,
        left: 0,
      },
      imageRect: {
        widthPx: widthPx,
        heightPx: heightPx,
      },
      frame: FrameShape.None,
      opacity: 1,
    },
  };

  return imageWidget;
};

/**
 * Get the absolute positioning style of an element
 */
export const parseElementAbsPosition = (el: HTMLElement): [number, number, number, number] => {
  const top = Math.abs(parseStrictNumber(el.style.top));
  const right = Math.abs(parseStrictNumber(el.style.right));
  const bottom = Math.abs(parseStrictNumber(el.style.bottom));
  const left = Math.abs(parseStrictNumber(el.style.left));

  return [top, right, bottom, left];
};

/**
 * Get the absolute position of the crop area
 */
export const cropToAbsolutePosition = (crop: Crop): [number, number, number, number] => {
  const top = -crop.top;
  const right = crop.right;
  const bottom = crop.bottom;
  const left = -crop.left;

  return [top, right, bottom, left];
};

/**
 * Get the transform translateX, translateY from an element
 */
export const getTranslateXY = (element: HTMLElement) => {
  const style = (window as any).getComputedStyle(element);
  const matrix = new DOMMatrixReadOnly(style.transform);

  return {
    translateX: matrix.m41,
    translateY: matrix.m42,
  };
};
