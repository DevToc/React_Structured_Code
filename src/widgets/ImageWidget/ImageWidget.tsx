import { ReactElement, useRef, useCallback, Suspense } from 'react';
import { Image, Skeleton, Box, Spinner } from '@chakra-ui/react';
import styled from '@emotion/styled';

import { useBoundingBox } from 'modules/Editor/components/BoundingBox/useBoundingBox';
import { useWidget, WidgetToolbar, WidgetBase, ReadOnlyWidgetBase, useEditor } from 'widgets/sdk';
import { CropHandler, CropData } from './Crop';
import { ImageWidgetData, ImageLayoutProps } from './ImageWidget.types';
import { useImageLoader, cropToAbsolutePosition } from './ImageWidget.helpers';
import { useImageResize } from './useImageResize';
import { CROP_BUTTON_ID } from './ImageWidget.config';
import { generateClipPathUrl, isSquareFrame } from './Frame/frame';
import { FrameShape } from './Frame/frame.types';
import { ClipPathFrameSvg } from './Frame/ClipPathFrameSvg';
import { useDynamicImport } from 'hooks/useDynamicImport';
import { WidgetComponentsMap } from 'widgets/components.lazy';

const ImageWidgetToolbarMenu = () => {
  const Component = useDynamicImport('imageWidgetToolbar', WidgetComponentsMap);

  return (
    <Suspense fallback={<Spinner />}>
      <Component />
    </Suspense>
  );
};

export const ImageWidget = (): ReactElement => {
  const { url, frame, rotateDeg, widgetId } = useWidget<ImageWidgetData>();
  const { isCropView, boundingBox } = useEditor();

  const { isLoading } = useImageLoader(url);
  const imageRectRef = useRef<HTMLDivElement>(null);
  const imageRef = useRef<HTMLImageElement>(null);

  const { onImageResize, onImageResizeEnd } = useImageResize({
    rotateDeg,
    widgetId,
    imageRectRef,
    imageRef,
    frame,
    boundingBox,
  });

  return (
    <WidgetBase onResize={onImageResize} onResizeEnd={onImageResizeEnd}>
      <WidgetToolbar>
        <ImageWidgetToolbarMenu />
      </WidgetToolbar>
      {isLoading ? (
        <Skeleton w='100%' h='100%' />
      ) : (
        <>{isCropView ? <CropContainer /> : <ImageLayout imageRectRef={imageRectRef} imageRef={imageRef} />}</>
      )}
    </WidgetBase>
  );
};

const CropContainer = () => {
  const { url, crop, frame, imageRect, widgetId, updateWidget, altText } = useWidget<ImageWidgetData>();
  const { widthPx: imageWidthPx, heightPx: imageHeightPx } = imageRect;

  const { zoom, disableEditorKeyboardShortcuts, enableEditorKeyboardShortcuts, disableCropView } = useEditor();
  const { widgetRefs, boundingBox } = useBoundingBox();
  const widgetBaseRef: HTMLElement = widgetRefs[widgetId]?.element;

  const onCropClose = useCallback(
    ({ crop, position }: CropData) => {
      enableEditorKeyboardShortcuts();

      if (!crop || !position) return;
      const widthPx = imageWidthPx - crop.left - crop.right;
      const heightPx = imageHeightPx - crop.top - crop.bottom;

      const { leftPx, topPx } = position;
      const widget = { widthPx, heightPx, leftPx, topPx, crop };

      updateWidget(widget);

      // Prevent widget bounding box to get out of sync when cropping with keyboard
      setTimeout(() => boundingBox.updateRect(), 0);
    },
    [updateWidget, enableEditorKeyboardShortcuts, imageHeightPx, imageWidthPx, boundingBox],
  );

  const closeCrop = useCallback(
    (e?: Event) => {
      if (e && e.target && (e.target as HTMLElement).id === CROP_BUTTON_ID) return;
      disableCropView();
    },
    [disableCropView],
  );

  return (
    <CropHandler
      isCropAspectRatioLocked={isSquareFrame(frame)}
      closeCrop={closeCrop}
      onCropOpen={disableEditorKeyboardShortcuts}
      onCropClose={onCropClose}
      imageWidth={imageWidthPx}
      imageHeight={imageHeightPx}
      altText={altText}
      url={url}
      crop={crop}
      zoom={zoom}
      widgetBaseRef={widgetBaseRef}
    />
  );
};

export const ReadOnlyImageWidget = (): ReactElement => {
  const { url } = useWidget<ImageWidgetData>();
  const { isLoading } = useImageLoader(url);

  return <ReadOnlyWidgetBase>{isLoading ? <Skeleton w='100%' h='100%' /> : <ImageLayout />}</ReadOnlyWidgetBase>;
};

const ImageLayout = ({ imageRectRef, imageRef }: ImageLayoutProps) => {
  const { url, crop, frame, imageRect, widgetId, opacity, altText, isDecorative } = useWidget<ImageWidgetData>();
  const { widthPx: w, heightPx: h } = imageRect;

  // Only apply alt text if specified or marked as decorative
  const altAttributes: { alt?: string } = {};
  if (altText || isDecorative) altAttributes['alt'] = isDecorative ? '' : altText;

  const [top, right, bottom, left] = cropToAbsolutePosition(crop);
  const hasClipPathFrame = !!FrameShape[frame] && frame !== FrameShape.None;
  const clipPathUrl = hasClipPathFrame ? generateClipPathUrl(widgetId) : '';

  return (
    <Box position='absolute' clipPath={clipPathUrl} overflow='hidden' w='100%' height='100%'>
      {hasClipPathFrame && <ClipPathFrameSvg frame={frame} id={widgetId} />}
      <div
        data-testid='image-rect'
        ref={imageRectRef}
        style={{ top, left, right, bottom, position: 'absolute', width: w, height: h, opacity }}
      >
        <ImageElement draggable='false' ref={imageRef} src={url} alt={altAttributes.alt} />
      </div>
    </Box>
  );
};

const ImageElement = styled(Image)`
  transform: translateZ(0px);
  backface-visibility: hidden;
  position: absolute;
  userselect: none;
  height: 100%;
  width: 100%;
  object-fit: fill;
`;
