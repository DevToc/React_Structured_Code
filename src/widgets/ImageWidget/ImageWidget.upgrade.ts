import { Widget } from 'types/widget.types';
import { ImageWidgetData } from './ImageWidget.types';

// the latest version of the widget
export const VERSION = 2;

class UpgradeV2 {
  version = 2;
  upgrade(widget: Widget) {
    const imageWidget = widget as ImageWidgetData;
    imageWidget.opacity = 1;
  }
}

// Add VersionControllers here that implement the WidgetVersionController interface
export const imageWidgetVersionControllers = [new UpgradeV2()];
