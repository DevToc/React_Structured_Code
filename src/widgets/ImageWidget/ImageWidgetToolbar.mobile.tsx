import { Suspense } from 'react';

import {
  Box,
  ButtonGroup,
  Drawer,
  DrawerOverlay,
  DrawerContent,
  DrawerHeader,
  DrawerBody,
  Flex,
  IconButton,
  Text,
  useDisclosure,
} from '@chakra-ui/react';

import { useWidget, useEditor } from 'widgets/sdk';
import { CROP_BUTTON_ID } from './ImageWidget.config';
import { ImageWidgetData } from './ImageWidget.types';
import { FRAME_SHAPE_LIST } from './Frame/frame.config';
import { FrameShape } from './Frame/frame.types';
import { isSquareFrame } from './Frame/frame';
import { AltTextMenu } from 'modules/Editor/components/Toolbar/AltTextMenu/AltTextMenu';
import { SliderNumberInput } from 'modules/common/components/Input/SliderNumberInput';

import { ReactComponent as CropIcon } from 'assets/icons/crop.svg';
import { ReactComponent as ImageFrameIcon } from 'assets/icons/rectangle.svg';
import { ReactComponent as CloseIcon } from 'assets/icons/close_circle.svg';
import { ReactComponent as OpacityIcon } from 'assets/icons/opacity.svg';

interface FrameOptionMenuProps {
  value: FrameShape;
  onChange: (v: FrameShape) => void;
}

const cropLabel = 'Crop';

export const ImageWidgetToolbarMenuMobile = () => {
  return (
    <Flex gap='8px'>
      <FrameOptionMenu />
      <CropMenu />

      <OpacityOptionWrapper />
      <AltTextMenuWrapper />
    </Flex>
  );
};

const OpacityOption = ({ value, onChange }: { value: number; onChange: (f: number) => void }) => {
  return (
    <SliderNumberInput
      title={'Opacity'}
      value={Math.trunc(value * 100)}
      onChange={onChange}
      suffix={'%'}
      sliderWidth={100}
      containerProps={{ p: 2, pt: 3 }}
    />
  );
};

const OpacityOptionWrapper = () => {
  const { opacity = 1, updateWidget } = useWidget<ImageWidgetData>();
  const { isOpen: isOpacityOpen, onToggle: toggleOpacity } = useDisclosure();

  const onChangeOpacity = (newValue: number) => updateWidget({ opacity: newValue / 100 });

  return (
    <>
      <Box
        onClick={() => toggleOpacity()}
        mr='5px'
        p='10px 10px'
        as='div'
        style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}
      >
        <IconButton mb='10px' size='sm' aria-label='Font Icon' icon={<OpacityIcon />} />
        <Box w='75px' as='div'>
          <Text fontSize={10} style={{ fontWeight: 700 }} align='center'>
            Opacity
          </Text>
        </Box>
      </Box>

      <Drawer placement='bottom' onClose={toggleOpacity} isOpen={isOpacityOpen}>
        <DrawerOverlay />
        <DrawerContent>
          <DrawerHeader borderBottomWidth='1px'>
            <Flex align='center' justify='space-between'>
              <p>Adjust Opacity</p>
              <CloseIcon onClick={toggleOpacity} />
            </Flex>
          </DrawerHeader>
          <DrawerBody>
            <Suspense fallback={<Box />}>
              <OpacityOption value={opacity} onChange={onChangeOpacity} />
            </Suspense>
          </DrawerBody>
        </DrawerContent>
      </Drawer>
    </>
  );
};

const AltTextMenuWrapper = () => {
  const { widgetId } = useWidget<ImageWidgetData>();

  return (
    <Box //
      mr='5px'
      p='10px 10px'
      as='div'
      style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}
    >
      <Box w='75px' as='div'>
        <AltTextMenu widgetId={widgetId} />
      </Box>
    </Box>
  );
};

const CropMenu = () => {
  const { isCropView, toggleCropView } = useEditor();

  return (
    <Box //
      mr='5px'
      p='10px 10px'
      as='div'
      style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}
    >
      <IconButton
        mb='10px'
        id={CROP_BUTTON_ID}
        onClick={toggleCropView}
        size='sm'
        aria-label={cropLabel}
        icon={<CropIcon style={{ pointerEvents: 'none' }} />}
        isActive={isCropView}
      />

      <Box w='75px' as='div'>
        <Text fontSize={10} style={{ fontWeight: 700 }} align='center'>
          Crop
        </Text>
      </Box>
    </Box>
  );
};

const FrameOptionMenu = () => {
  const { updateWidget, widthPx, heightPx, crop, imageRect, frame: value } = useWidget<ImageWidgetData>();
  const { boundingBox } = useEditor();
  const { isOpen: isCropShapeOpen, onToggle: toggleCropShape } = useDisclosure();

  const onChangeFrame = (frame: FrameShape) => {
    if (isSquareFrame(frame)) {
      const isWidthLarger = widthPx >= heightPx;
      const squareSidePx = Math.min(widthPx, heightPx);

      let newCrop = { ...crop };
      // Update the crop position to the new square size
      if (isWidthLarger) newCrop.right = imageRect.widthPx - squareSidePx - crop.left;
      else newCrop.bottom = imageRect.heightPx - squareSidePx - crop.top;

      updateWidget({ widthPx: squareSidePx, heightPx: squareSidePx, crop: newCrop, frame });

      // Prevent widget bounding box to get out of sync if the widget is resized
      return setTimeout(() => boundingBox.updateRect(), 0);
    }

    return updateWidget({ frame });
  };

  return (
    <>
      <Box
        onClick={() => toggleCropShape()}
        mr='5px'
        p='10px 10px'
        as='div'
        style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}
      >
        <IconButton
          mb='10px'
          size='sm'
          aria-label='Font Icon'
          icon={<ImageFrameIcon style={{ pointerEvents: 'none' }} />}
        />
        <Box w='75px' as='div'>
          <Text fontSize={10} style={{ fontWeight: 700 }} align='center'>
            Crop Shape
          </Text>
        </Box>
      </Box>

      <Drawer placement='bottom' onClose={toggleCropShape} isOpen={isCropShapeOpen}>
        <DrawerOverlay />
        <DrawerContent>
          <DrawerHeader borderBottomWidth='1px'>
            <Flex align='center' justify='space-between'>
              <p>Select Shape</p>
              <CloseIcon onClick={toggleCropShape} />
            </Flex>
          </DrawerHeader>
          <DrawerBody>
            <Suspense fallback={<Box />}>
              <FrameOptionMenuMobile value={value} onChange={onChangeFrame} />
            </Suspense>
          </DrawerBody>
        </DrawerContent>
      </Drawer>
    </>
  );
};

const FrameOptionMenuMobile = ({ value = FrameShape.None, onChange }: FrameOptionMenuProps) => {
  return (
    <Box p={2}>
      <ButtonGroup gap='8px' variant='outline' spacing='0' display='flex' flexWrap='wrap' size='sm'>
        {FRAME_SHAPE_LIST.map((frameShape) => {
          const { label, value: frameValue, icon: Icon } = frameShape;
          const isActive = value === frameValue;
          const handleClick = () => onChange(frameValue);

          return (
            <IconButton isActive={isActive} onClick={handleClick} aria-label={label} key={frameValue} icon={<Icon />} />
          );
        })}
      </ButtonGroup>
    </Box>
  );
};
