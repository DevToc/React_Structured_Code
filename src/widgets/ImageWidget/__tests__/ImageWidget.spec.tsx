import { screen } from '@testing-library/react';
import { renderWidget } from 'utils/test-utils.test';
import { ImageWidget } from '../ImageWidget';
import { generateDefaultData } from '../ImageWidget.helpers';

const DEFAULT_ALT_TEXT = 'I am alt text.';

describe('widgets/ImageWidget', () => {
  it('should render skeleton correctly', () => {
    const widgetId = '007.widget-1';
    const mockSize = { width: 100, height: 100 };
    const mockUrl = 'cat.png';
    const widget = { widgetId, ...generateDefaultData(mockUrl, mockSize, mockSize) };

    const { asFragment } = renderWidget(<ImageWidget />, { widget });

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render image correctly', async () => {
    const widgetId = '007.widget-1';
    const mockSize = { width: 100, height: 100 };
    const widget = { widgetId, ...generateDefaultData('', mockSize, mockSize) };

    const { asFragment } = renderWidget(<ImageWidget />, { widget });

    await screen.findByTestId('image-rect');
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render alt text', () => {
    const widgetId = '007.widget-1';
    const mockSize = { width: 100, height: 100 };
    const defaultData = generateDefaultData('', mockSize, mockSize);

    const widget = {
      widgetId,
      ...defaultData,
      widgetData: {
        ...defaultData.widgetData,
        altText: DEFAULT_ALT_TEXT,
        isDecorative: false,
      },
    };

    renderWidget(<ImageWidget />, { widget });
    expect(screen.getByAltText(DEFAULT_ALT_TEXT)).toBeInTheDocument();
  });

  it('should render as decorative', () => {
    const widgetId = '007.widget-1';
    const mockSize = { width: 100, height: 100 };
    const defaultData = generateDefaultData('', mockSize, mockSize);

    const widget = {
      widgetId,
      ...defaultData,
      widgetData: {
        ...defaultData.widgetData,
        altText: null,
        isDecorative: true,
      },
    };

    renderWidget(<ImageWidget />, { widget });

    expect(screen.getByAltText('')).toBeInTheDocument();
  });

  it('should render the opacity as expected', async () => {
    const widgetId = '007.widget-1';
    const mockSize = { width: 100, height: 100 };
    const defaultData = generateDefaultData('', mockSize, mockSize);

    const widget = {
      widgetId,
      ...defaultData,
      widgetData: {
        ...defaultData.widgetData,
        altText: null,
        opacity: 1,
      },
    };

    renderWidget(<ImageWidget />, { widget });
    const imageRect = await screen.findByTestId('image-rect');

    expect(imageRect.style).toHaveProperty('opacity', '1');
  });
});
