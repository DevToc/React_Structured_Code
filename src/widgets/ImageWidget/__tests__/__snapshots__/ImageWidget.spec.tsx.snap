// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`widgets/ImageWidget should render image correctly 1`] = `
<DocumentFragment>
  <div
    aria-describedby="007.widget-1-description"
    aria-label="Image widget"
    class="widget-container widget-selecto-target  css-1eee8vp"
    data-testid="widgetbase-007.widget-1"
    id="007.widget-1"
    style="display: block; position: absolute; top: 0px; left: 0px; width: 50px; height: 50px; transform: rotate(0deg); z-index: 0; cursor: move;"
    tabindex="-1"
  >
    <div
      class="css-7cxduo"
    >
      <div
        data-testid="image-rect"
        style="top: 0px; left: 0px; right: 0px; bottom: 0px; position: absolute; width: 50px; height: 50px; opacity: 1;"
      >
        <img
          class="css-b6n3l3"
          draggable="false"
          src=""
        />
      </div>
    </div>
    <span
      class="css-f8n5zr"
    >
      <p
        class="chakra-text css-0"
        id="007.widget-1-description"
      />
    </span>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`widgets/ImageWidget should render skeleton correctly 1`] = `
<DocumentFragment>
  <div
    aria-describedby="007.widget-1-description"
    aria-label="Image widget"
    class="widget-container widget-selecto-target  css-1eee8vp"
    data-testid="widgetbase-007.widget-1"
    id="007.widget-1"
    style="display: block; position: absolute; top: 0px; left: 0px; width: 50px; height: 50px; transform: rotate(0deg); z-index: 0; cursor: move;"
    tabindex="-1"
  >
    <div
      class="chakra-skeleton css-jhptx"
    />
    <span
      class="css-f8n5zr"
    >
      <p
        class="chakra-text css-0"
        id="007.widget-1-description"
      />
    </span>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
